streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoin.java

```diff
@@ -183,15 +195,27 @@ public void process(final K key, final V1 value) {
 
         @SuppressWarnings("unchecked")
         private void emitNonJoinedOuterRecords(final WindowStore<KeyAndJoinSide<K>, LeftOrRightValue> store) {
+            if (sharedTimeTracker.minTime >= sharedTimeTracker.streamTime - joinAfterMs - joinGraceMs) {
+                return;
+            }
+            if (context.currentSystemTimeMs() < sharedTimeTracker.nextTimeToEmit) {
```

Comment:
Could you add a comment why we're using the system time instead of stream time to emit the results? I think I know why, but it would be good to have it here for future understanding.

Reply:
Btw, this won't get deterministic results, right? If I re-run the join from the beginning, then I might get the expired records only after 50ms of my time has passed, not by the stream time. Is that desirable?

Reply:
I think the result is still deterministic, as the result will always be the same records; only the output order might change, what is no issue IMHO.